File: backend/modules/petclinic/app/petclinic/api/pets/updatePetDetail/UpdatePetDetailService.java
package petclinic.api.pets.updatePetDetail;

import akka.actor.ActorRef;
import akka.pattern.AskSupport;
import akka.pattern.PatternsCS;
import petclinic.api.model.OwnerId;
import petclinic.api.model.PetType;
import petclinic.api.pettype.getPetType.GetPetTypeService;
import petclinic.model.OwnerMessages;

import javax.inject.Inject;
import javax.inject.Named;
import java.time.LocalDate;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import static petclinic.api.Constants.TIMEOUT;

public class UpdatePetDetailService implements AskSupport {

    @Inject
    @Named("owner")
    private ActorRef owner;

    @Inject
    private GetPetTypeService petTypeService;

    public CompletionStage<Optional<UpdatePetDetailApiResponse>> updatePetDetails(OwnerId ownerId, String petId, String name, LocalDate birthDay, String petTypeId) {
        return petTypeService.getPetType(petTypeId).thenCompose(maybePetType -> {
            if (!maybePetType.isPresent()) {
                return CompletableFuture.completedFuture(Optional.<UpdatePetDetailApiResponse>empty());
            } else {
                PetType petType = maybePetType.get();
                return PatternsCS.ask(owner, OwnerMessages.updatePetDetails(ownerId.id(), petId, name, birthDay, petType), TIMEOUT)
                        .thenApply(obj -> (OwnerMessages.UpdatePetDetailsResponse) obj)
                        .thenApply(obj -> {
                            if (obj instanceof OwnerMessages.OwnerNotFound) {
                                return Optional.empty();
                            } else if (obj instanceof OwnerMessages.PetNotFound) {
                                return Optional.empty();
                            } else if (obj instanceof OwnerMessages.PetDetailsUpdatedResponse) {
                                return Optional.of(UpdatePetDetailApiResponse.apply(((OwnerMessages.PetDetailsUpdatedResponse) obj).petId()));
                            } else {
                                throw new IllegalStateException("unsupported state");
                            }
                        });
            }
        });
    }

}

File: backend/modules/petclinic/app/petclinic/api/pets/updatePetDetail/UpdatePetDetailService.scala
package petclinic.api.pets.updatePetDetail

import java.time.LocalDate

import akka.actor.ActorRef
import javax.inject.{Inject, Named}
import petclinic.api.Constants
import petclinic.api.model.OwnerId
import petclinic.api.pettype.getPetType.GetPetTypeService
import petclinic.model.owners.owner.Protocol._

import scala.concurrent.{ExecutionContext, Future}

class UpdatePetDetailService @Inject() (@Named(petclinic.api.Constants.OWNER_AGGREGATE_ACTOR_NAME) actorRef: ActorRef, getPetTypeService: GetPetTypeService)(implicit ec: ExecutionContext) {

  def updatePetDetails(ownerId: OwnerId, petId: String, name: String, birthDay: LocalDate, petTypeId: String): Future[Option[UpdatePetDetailApiResponse]] = {
    getPetTypeService.getPetType(petTypeId)
      .flatMap(maybePetType => {
      maybePetType.map(pt => {
        akka.pattern.Patterns.ask(actorRef, UpdatePetDetailsCommand(ownerId.id, petId, name, birthDay, pt), Constants.TIMEOUT)
          .mapTo[UpdatePetDetailsResponse]
          .map {
            case OwnerNotFoundUpdatePetDetailsResponse => None
            case PetNotFoundUpdatePetDetailsResponse => None
            case PetDetailsUpdatedResponse(petIdReceived) => Some(UpdatePetDetailApiResponse(petIdReceived))
          }
      }).getOrElse(Future.successful(None))

    })
  }

}
